src/messages/MMDSOp.h

```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include "msg/Message.h"
+
+class MMDSOp: public SafeMessage {
```

Comment:
how does `MMDSOp` ensure that its derived classes are versioned?

Reply:
It checks if the inter-mds message classes are [inherited from MMDSOp or not](https://github.com/ceph/ceph/pull/31726/files#diff-ea4df8dd62d8f79137676b9f21d58633R994). If not then assert. That's the solution suggested by @batrick [here](https://tracker.ceph.com/issues/41565).

Reply:
i think my question is more related to "versioned". in your change `MMDSOp` is no more than an alias of `SafeMessage` with its own RTTI information. it does **not** enforce versioning.

as the title of this PR is "detect unversioned MDS-MDS messages" and the tracker ticket is named "detect MDS<->MDS messages that are not versioned".

Reply:
There isn't really a good way AFAIK to enforce versioning for MDS-MDS messages. The approach I've suggested is a very low barrier which requires messages between MDS to at least inherit `MMDSOp`. Do you have a better suggestion @tchaikov?

Reply:
@batrick and @joscollin couple questions first, do we plan to reject an MDS instance whose version is not compatible with the lowest-compatible version, a.k.a. minimum required version? do we add all MDS-MDS messages to corpus ?

if both answers are yes, i think MDS-MDS messages are already versioned.

Reply:
> do we plan to reject an MDS instance whose version is not compatible with the lowest-compatible version, a.k.a. minimum required version?

The eventual goal is to support rolling upgrades. The MDSs should be able to communicate with an older MDS long enough to finish the upgrade. Any new features/fixes should not require incompatible message versions. New features should be gated in such a way that either it doesn't require an older MDS to participate or there is a flag to enable it. So I believe the answer to your question is no.

> if both answers are yes, i think MDS-MDS messages are already versioned.

The issue this PR is trying to address is to prevent a new message from being introduced that is not versioned. At the very least, the author of the new message would need to inherit from `MMDSOp`. To drive that home, @joscollin please add a big comment above `MMDSOp`'s definition explaining what `MMDSOp` is: that any MDS-MDS messages _must_ be versioned and `MMDSOp` exists to make you remember that.

Reply:
> New features should be gated in such a way that either it doesn't require an older MDS to participate or there is a flag to enable it. So I believe the answer to your question is no.

actually, i interpret your answer as "yes". because, it sounds exactly like what we have when handling the "require-osd-release" command and `OSDMap::require_osd_release` . as OSD also supports rolling upgrades. and new OSD is able to talk to older OSD, but if the peer OSD is too old, the monitor will reject it when it tries to join the cluster. and "osd require-osd-release" raises the bar higher once all monitors and OSDs in cluster are newer than the given release.

i don't think MDS maintains would like to be backward compatible with ancient MDS so they can join and talk with the latest MDS. but i think MDS of release *n* should be able to talk to at least MDS release *n-2*.

that's why i think it'd be important to have MDS messages collected in corpus to make sure the decoders of "n" release are able to to read the messages emitted by "n-2" release.

what confuses me is that `MMDSOp` does not add any runtime/compile time error for unversioned MDS messages. but if we add the MDS messages to corpus, we will be able to test the backward compatibility of the dencoder of them.

Reply:
You raise a good point Kefu. I think we still want this patch though as it's useful for identifying what messages are passed between MDS (for new developers). The check also ensures new message types passed between MDS must also inherit MMDSOp. I do also foresee us using `MMDSOp` to regularly pass some information between MDSs in the future, like ping epochs (for `fs top`).

The next step for @joscollin in this series will be to collect messages for the corpus.